Add "protocols" to define metaprogramming interfaces

nilsvu opened the pull request:
## Proposed changes

This PR implements what I suggested in #1349. It adds a dev guide to the documentation that explains a lot of the reasoning and implementation, so I don't repeat it here in the description. This PR is a proposal to introduce this new concept, so I'm happy for any feedback!

<s>As an example, I add protocols for analytic solutions and analytic data in this PR. It replaces the current `MarkAsAnalyticSolution` struct (which is basically a protocol already) with a more structured approach.</s>

- [x] Depends on #1793

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

moxcodes commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Test_Protocols.cpp:
```diff
@@ -0,0 +1,145 @@
+      const tnsr::I<DataVector, Dim>& x, const double t,
+      const PrecomputedData& precomputed_data,
+      tmpl::list<SomeField> /*meta*/) noexcept {
+    return {Scalar<DataVector>{precomputed_data * t * get<0>(x)}};
```
This might not need definitions of the functions in order to pass the protocol checks (because they're all compile-time)

nilsvu replied:
Indeed it doesn't! Everything but the function declaration can be removed. What do you think, should the test provide the implementation nonetheless so it appears in the documentation and gives an example, or should it be reduced to only the function declaration?

moxcodes reviewed (commented):
Group code review (start) with @fmahebert and @nilsleiffischer

moxcodes commented on src/PointwiseFunctions/AnalyticData/Protocols.hpp:
```diff
@@ -0,0 +1,101 @@
+ * \ingroup ProtocolsGroup
+ * \brief A analytic initial data to a system of evolution equations.
+ *
+ * Requires the class has this metaprogramming interface:
```
@nilsleiffischer : should say that the class must have these static member variables

moxcodes commented on src/PointwiseFunctions/AnalyticData/Protocols.hpp:
```diff
@@ -0,0 +1,101 @@
+ */
+struct AnalyticData {
+ private:
+  CREATE_HAS(volume_dim)
```
Probably best to factor out the common code between the protocols.

moxcodes commented on src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp:
```diff
@@ -87,10 +90,12 @@ template <size_t Dim, typename InitialData>
 struct EvolutionMetavars {
   static constexpr size_t volume_dim = Dim;
   using initial_data = InitialData;
+  static constexpr bool has_analytic_solution =
```
General comment: Consider the possibility of performing the "rigorous" check at the point of the `conforms_to` invocation depending on the compile flags (e.g. debug builds).

nilsvu replied:
Notes from our discussion:
- We should probably have a cmake option that enables these rigorous protocol conformance checks in `conforms_to_v`.
- We can try enabling the option for debug builds. It is not clear to us how much this will hurt compile times once we have many `conforms_to_v` checks in the code. We should never be discouraged from adding more of these checks!
- We considered the possibility that having the rigorous check in debug builds would eliminate the need to add any protocol conformance checks in unit tests. But thinking about this more I disagree:
  - In any case we need to check in unit tests that the class derives off the protocol, i.e. indicates in conforms to it. So we could just add `conforms_to_v` checks in the unit tests. But I think it's more flexible and future-proof (and doesn't require any more effort in code review) to have the dedicated `test_protocol_conformance` metafunction in unit tests.

So in conclusion I suggest to keep everything as-is and add an optional cmake flag that changes the behavior of `conforms_to_v`.

moxcodes commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagneticFieldLoop.hpp:
```diff
@@ -197,9 +198,6 @@ class MagneticFieldLoop : public evolution::protocols::AnalyticData {
   tuples::TaggedTuple<Tags...> variables(
       const tnsr::I<DataType, 3, Frame::Inertial>& x,
       tmpl::list<Tags...> /*meta*/) const noexcept {
-    static_assert(sizeof...(Tags) > 1,
```
Probably replace all of these with a better test involving the list of supported tags.

nilsvu replied:
... and add the list of supported tags to the analytic solution compile-time interface.

moxcodes reviewed (commented):
Continued group code review with @nilsleiffischer, @fmahebert , with @moxcodes primary.
We didn't get through the first main commit with all of the docs

nilsvu commented:
Thanks for your feedback so far @moxcodes and @fmahebert. I think we now have a pretty good idea how protocols can be used in practice, so I removed most of the code that refactors analytic solutions from this PR to make it easier to review. I only left an example each for evolution and elliptic solutions. I'll make the other solutions conform to the protocols in a follow-up PR that also removes the `MarkAs*` structs.

An important insight that I think @fmahebert pointed out is that we should probably never make protocol _templates_, but instead include the parameters in the interface that the protocol defines.

nilsdeppe commented:
Can you expand on the second paragraph with some code examples? I don't understand what it means.

nilsvu commented:
@nilsdeppe I pushed a fixup that adds a section about it to the documentation. Once GitHub CI is done, you can try the new artifact feature to take a look at the docs :)

moxcodes commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -938,6 +938,36 @@ using is_callable_t = typename is_callable<T, Args...>::type;
+// alias type at all. If someone has that many pointers to a thing that isn't
+// useful, it's their fault...
+#define CREATE_HAS_TYPE_ALIAS(ALIAS_NAME)                                     \
+  template <typename CheckingType, typename AliasType = NoSuchType*****,      \
```
Can you just create a new useless struct instead of putting five pointers on this thing? I'd find that more elegant.

moxcodes commented on src/PointwiseFunctions/AnalyticSolutions/Protocols.hpp:
```diff
@@ -13,6 +13,54 @@
+template <typename ConformingType, bool WithTime, typename TagsList>
+struct IsVariablesCallable;
+template <typename ConformingType, typename TagsList>
+struct IsVariablesCallable<ConformingType, true, TagsList>
```
it'd be better to add a code comment explaining that this bool is for the time variable that might or might not be present for analytic solutions.

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -0,0 +1,94 @@
+developer who consumes the protocol by writing classes that conform to it knows
+exactly what needs to be implemented.
+
+We typically define protocols in a file named `Protocols.hpp` and within a
```
Clariify that the `hpp` should be placed in the directory associated with the utilities that depend on the interface conformance.

moxcodes commented on tests/Unit/ProtocolTestHelpers.hpp:
```diff
@@ -0,0 +1,29 @@
+
+}  // namespace detail
+
+template <typename ConformingType, typename Protocol>
```
this should have rendered documentation, and in order for the dox to render you will also have to whitelist it in the `/docs`

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -0,0 +1,94 @@
+
+\snippet Utilities/Test_ProtocolHelpers.cpp named_protocol
+
+It uses the `CREATE_IS_CALLABLE` macro from `Utilities/TypeTraits.hpp` to make
```
try to get the `CREATE_IS_CALLABLE` to link (e.g. add parens or use \ref)

moxcodes commented on src/Utilities/ProtocolHelpers.hpp:
```diff
@@ -0,0 +1,23 @@
+// will not match private base classes (unlike std::is_base_of)
+
+// @{
+/// Checks if the `ConformingType` conforms to the `Protocol`.
```
if you put this in a group it will render.

moxcodes commented on tests/Unit/ProtocolTestHelpers.hpp:
```diff
@@ -0,0 +1,29 @@
+namespace detail {
+
+template <typename ConformingType, typename Protocol>
+struct TestProtocolConformanceImpl : std::true_type {
```
the value for this struct can just be the `is_conforming_v` from the second assert, which would then be checked below.

moxcodes reviewed (commented):
Group code review with @nilsleiffischer , @moxcodes primary.

nilsvu commented:
@sxs-collaboration/spectre-core-devs Could you review this PR or assign a reviewer? I stripped this PR of all the refactoring code, so it just adds the protocols utility and example protocols for analytic solutions/data. The changes are non-breaking, so we can gradually adopt protocols without needing to refactor existing code, as we have discussed in the code meeting.
I’ll push fixups for @moxcodes review asap.

nilsvu committed:
Add CMake option to enable rigorous protocol conformance checks

nilsvu commented:
@moxcodes @nilsdeppe I updated this PR and removed everything but the protocol utility, tests and documentation so that it's easier to review. I'd like to make use of protocols in a few upcoming PRs, and we have already some code that really should be protocols (e.g. the DataBox tag tests that @kidder has been writing), so I made this PR as small as possible. Could you take a look when you get the chance?

kidder commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -858,54 +859,67 @@ using is_callable_t = typename is_callable<T, Args...>::type;
+  };
+
+#define CREATE_IS_CALLABLE_R_V(METHOD_NAME)                     \
+  CREATE_IS_CALLABLE_R(METHOD_NAME)                             \
```
don't call this in the macro, make the user call two macros

nilsvu committed:
Split CREATE_IS_CALLABLE into separate macros

The macros follow the standard library's naming convention. We split
them into individual macros to avoid "unused variable" compiler warnings.

nilsvu commented:
@moxcodes @nilsdeppe pushed a fixup with the changes we discussed in the call yesterday. Please take a look :)

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -22,7 +26,8 @@ protocol through metaprogramming or documentation, and unit tests that enforce
 You should define a protocol when you need a template parameter to conform to an
-interface. Here is an example of a protocol that is taken from the
+interface. Protocols are implemented as unary type traits. Here is an example of
+a protocol that is taken from the
```
I'd use the term `adapted from` rather than `taken from` because the example isn't identical due to language differences.

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -32,12 +37,23 @@ For example, the following class conforms to the protocol we just defined:
 
 \snippet Utilities/Test_ProtocolHelpers.cpp named_conformance
 
+The class indicates it conforms to the protocol by (publicly) deriving off
```
I'd prefer "Inherit from" rather than "deriving off"

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -32,12 +37,23 @@ For example, the following class conforms to the protocol we just defined:
 
-So now you can write code that relies on the interface defined by the protocol:
+Note that checking for protocol conformance is cheap, because the `conforms_to`
+metafunction only checks if the class _indicates_ it conforms to the protocol.
```
it would be helpful to insert "_indicates_ it conforms to the protocol via the above inheritance" just to reiterate and make the statement totally clear.

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -48,26 +64,40 @@ has explicitly defined (and documented!) the interface they expect. And the
-part of your protocol instead. The reason for this guideline is that protocols
+When you author a new protocol, keep in mind that protocols must be unary type
+traits. This means they take a single template parameter (typically named
+`ConformingType`) and derive off `std::true_type` or `std::false_type` depending
```
"inherit from"

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -48,26 +64,40 @@ has explicitly defined (and documented!) the interface they expect. And the
+Occasionally, You might be tempted to add additional template parameters to the
+protocol. In those situations, make the additional parameters part of your
+protocol instead. The reason for this guideline is that protocols
 will always be used as base classes. Therefore, any template parameters of the
```
"be used as base classes" -> "be used as unary type traits when inheriting from `ConformsTo<Protocol>`"

moxcodes commented on docs/DevGuide/Protocols.md:
```diff
@@ -48,26 +64,40 @@ has explicitly defined (and documented!) the interface they expect. And the
 \snippet Utilities/Test_ProtocolHelpers.cpp test_protocol_conformance
 
-## Protocol authors: Protocols should not be templates
+# Protocol authors: Protocols must be unary type traits {#protocols_author}
```
I'm not sure I completely agree with the suggestions in this section anymore now that the protocols are unary type traits. If we want to pass an extra template parameter, we could for instance define a struct with a metafunction within:
```
template <typename NameType>
struct is_name_callable {
template <typename ToCheck>
using check_conformance = std::conditional_t< (checks) >;
};
```
Then the inheritance can be `ConformsTo<is_name_callable<ThisNameType>::check_conformance>` and similar in the conformance checking metafunction. This gives the same flexibility as extra template parameters without disrupting the interface you need to supply to the checking metafunctions. What do you think?

nilsvu replied:
- In your example the `check_conformance` is the protocol, so it's still a unary type trait.
- I think the statement still holds that all potential template parameters for a protocol must be known to the class that conforms to it (via the `ConformsTo` inheritance) and therefore should be made part of the protocol. In your example it's `ThisNameType`. Instead of passing `ThisNameType` to the struct that "generates" the `check_conformance` metafunction, you can just make `NameType` part of the interface (as described in the docs).
- This guideline also encourages classes to publicly expose their template parameters as type aliases, which is a good thing to do. There's no reason that I can think of why the class in you example shouldn't expose its `ThisNameType` template parameter as a type alias.

moxcodes replied:
Yeah, alright. I'm convinced that this is a fine strategy for almost all use cases. It's probably also true that anyone who finds a complex enough edge-case to want the nested type templates will also be able to figure that solution out for themselves.

moxcodes commented on src/Utilities/ProtocolHelpers.hpp:
```diff
@@ -7,11 +7,22 @@
+/*!
+ * \brief Indicate a class conforms to the `Protocol`.
+ *
+ * (Publicly) derive classes off this class to indicate they conform to the
```
"inherit classes from"

moxcodes commented on src/Utilities/ProtocolHelpers.hpp:
```diff
@@ -20,21 +31,22 @@
  */
 #ifdef SPECTRE_ALWAYS_CHECK_PROTOCOL_CONFORMANCE
-template <typename ConformingType, typename Protocol>
+template <typename ConformingType, template <class> class Protocol>
```
the template template can be `template <typename> class` (both don't have to be class), though I don't think it has any functional difference, so I'll leave it up to you for whether you want to change it.

nilsvu replied:
I chose `class` for both to avoid confusion why one is `class` and the other is `typename` 🤷‍♂️ The restriction will be lifted in cpp17

nilsdeppe commented on docs/DevGuide/Protocols.md:
```diff
@@ -48,26 +64,40 @@ has explicitly defined (and documented!) the interface they expect. And the
+the protocols above for testing the existence and return type of a member
+function.
+
+Occasionally, You might be tempted to add additional template parameters to the
```
[nitpick] `You` -> `you` (lower case the first letter)

nilsdeppe commented on src/Utilities/ProtocolHelpers.hpp:
```diff
@@ -7,11 +7,22 @@
 
 #include "Utilities/TypeTraits.hpp"
 
+/*!
```
I would like to see a `protocols` or `proto` namespace. I've often found our pollution of the global namespace a bit annoying (e.g. domain...)

nilsvu replied:
Yes my intention is that protocols are placed in `protocols` namespaces (just like our `Tags` namespaces). This is written at the end of the overview section in the docs.

The `Utilities/ProtocolHelpers.hpp` defines utility metafunctions that are not protocols. I suppose we could place them in a top-level `protocols` namespace, but that could be misleading. E.g. the following does not read very nicely:
```cpp
struct Person : protocols::ConformsTo<protocols::Named> {};

static_assert(protocols::conforms_to<Person, protocols::Named>, "...");
```
A lot of the `Utilities` are in the global namespace, and that's probably fine because they supply general functionality that could be in the standard library, as opposed to things in `Domain` that should really be in a namespace. So the code above then reads:
```cpp
struct Person : ConformsTo<protocols::Named> {};

static_assert(conforms_to<Person, protocols::Named>, "...");
```
What do you think?

nilsdeppe replied:
I think this needs to go in a namespace... Personally I think we should've written all of spectre inside a `spectre` namespace with appropriate nested namespaces but here we are...

nilsvu replied:
Ok, how about something like a top-level `prtcl` namespace? Or use `tt`?

nilsvu replied:
or `util`?

nilsdeppe replied:
I'd just go with protocols or protos or proto. I don't love `util` because that's rather vague. I would find it easier if all the protocols stuff is in a `protocols` namespace, then I don't have to think about which namespace things are in when. The other option is to put all the protocols into the `tt` namespace, since now they really are just type traits

nilsvu replied:
👍 I moved the protocol helpers into the top-level `tt` namespace. Please check the fixup

nilsdeppe commented on docs/DevGuide/Protocols.md:
```diff
@@ -22,7 +26,8 @@ protocol through metaprogramming or documentation, and unit tests that enforce
 conformance to these protocols.
```
I think it might be good to emphasize already after the first sentence in the paragraph that protocols is not really like concepts and is trying to solve a different problem, but that the way protocols are implemented, they can easily be used with concepts. I think it would be good to have a short section comparing what concepts solves to what protocols solves, and how they complement each other.

nilsdeppe reviewed (commented):
I'm happy for you to squash whenever @moxcodes is happy

nilsvu commented:
Pushed fixups @moxcodes @nilsdeppe. Please take a look :)

moxcodes commented:
I'm happy for you to squash.

nilsdeppe commented on docs/DevGuide/Protocols.md:
```diff
@@ -143,3 +137,17 @@ example implementation of a class that conforms to the protocol, and the
+
+# Protocols and C++20 "Constraints and concepts" {#protocols_and_constraints}
+
+A feature related to protocols is proposed for C++20 and goes under the name of
```
While squashing, this is officially in C++20 now as of last week or so. `proposed for` -> `in`

nilsdeppe reviewed (commented):
I'm happy for you to squash, including the last minor comment.

nilsvu commented:
This is rebased in squashed. Thanks for reviewing @nilsdeppe @moxcodes :)

kidder commented:
@wthrowe do you have any general comments about this PR?  It's already been thoroughly reviewed...

wthrowe commented:
I read through the documentation.  This looks reasonable.

nilsvu commented:
Rebased to resolve a conflict with #1937

kidder commented:
@nilsleiffischer this is ready if you can get GitHub actions to work

nilsvu commented:
@kidder could you restart the CI build? I can only trigger a rebuild by force-pushing

kidder commented:
@nilsleiffischer when I restart the checks everything fails...

nilsvu committed:
Add CREATE_HAS_TYPE_ALIAS macro for type traits

nilsvu committed:
Add protocols utility and documentation

Protocols are a concept we are introducing to define
metaprogramming interfaces.
